lighthouse-core/scripts/i18n/cull-obsolete-lhl-messages.js

```diff
@@ -0,0 +1,152 @@
+  console.log(`Removing message\n\t'${messageId}'\nfrom translations: ${reason}.`);
+  alreadyLoggedCulls.add(messageId);
+}
+
+/**
+ * Returns a copy of `localeLhl` with only messages matching those from the golden locale.
+ * `goldenLocaleArgumentIds` values are assumed to be sorted.
+ * @param {Record<string, Array<string>>} goldenLocaleArgumentIds
+ * @param {LhlMessages} localeLhl
+ * @param {Set<string>} alreadyLoggedCulls Set of culls that have been logged and shouldn't be logged again.
+ * @return {LhlMessages}
+ */
+function cullLocale(goldenLocaleArgumentIds, localeLhl, alreadyLoggedCulls) {
+  /** @type {LhlMessages} */
+  const remainingMessages = {};
+
+  for (const [messageId, {message}] of Object.entries(localeLhl)) {
+    const goldenArgumentIds = goldenLocaleArgumentIds[messageId];
+    if (!goldenArgumentIds) {
+      logRemoval(alreadyLoggedCulls, messageId, 'it is no longer found in Lighthouse');
+      continue;
+    }
+
+    if (!equalArguments(goldenArgumentIds, message)) {
+      logRemoval(alreadyLoggedCulls, messageId,
+          'its ICU arguments don\'t match the current version of the message');
+      continue;
+    }
+
+    remainingMessages[messageId] = {message};
+  }
+
+  return remainingMessages;
+}
+
+/**
+ * Returns a copy of `goldenLhl` with the messages replaced with a sorted list of
+ * argument ids found in each message.
+ * @param {LhlMessages} goldenLhl
+ * @return {Record<string, Array<string>>}
+ */
+function getGoldenLocaleArgumentIds(goldenLhl) {
+  /** @type {Record<string, Array<string>>} */
+  const goldenLocaleArgumentIds = {};
+
+  for (const [messageId, {message}] of Object.entries(goldenLhl)) {
+    const parsedMessage = MessageParser.parse(message);
+    const goldenArgumentElements = collectAllCustomElementsFromICU(parsedMessage.elements);
+    const goldenArgumentIds = [...goldenArgumentElements.keys()].sort();
+
+    goldenLocaleArgumentIds[messageId] = goldenArgumentIds;
+  }
+
+  return goldenLocaleArgumentIds;
+}
+
+/**
+ * For every locale LHL file, remove any messages that don't have a matching
+ * message in `en-US.json`. There is a matching golden message if:
+ * - there is a golden message with the same message id and
+ * - that message has the same ICU arguments (by count and argument ids).
+ *
+ * If a new `en-US.json` message is sufficiently different so that existing
+ * translations should no longer be used, it's up to the author to remove them
+ * (e.g. by picking a new message id).
+ */
+function cullObsoleteLhlMessages() {
+  const goldenLhl = require('../../lib/i18n/locales/en-US.json');
+  const goldenLocaleArgumentIds = getGoldenLocaleArgumentIds(goldenLhl);
+
+  // Find all locale files, ignoring self-generated en-US, en-XL, and ctc files.
+  const ignore = [
+    '**/.ctc.json',
+    '**/en-US.json',
+    '**/en-XL.json',
+  ];
+  const globPattern = 'lighthouse-core/lib/i18n/locales/**/+([-a-zA-Z0-9]).json';
+  const lhRoot = `${__dirname}/../../../`;
```

Comment:
is there a particular reason it's not `path.join`'d?

Reply:
at least for the glob pattern, `/` should always be used https://github.com/isaacs/node-glob#windows

for `cwd` .... what happens if you do `path.resolve('C:/some/file.js')` in windows? https://github.com/isaacs/node-glob/blob/master/common.js#L97

Reply:
> is there a particular reason it's not path.join'd?

cause this line is all it's doing :)

> for `cwd` .... what happens if you do `path.resolve('C:/some/file.js')` in windows? https://github.com/isaacs/node-glob/blob/master/common.js#L97 (btw, a few lines later they normalize the path)

it turns `'C:/some/file.js'` into `'C:\some\file.js'`, which then, yeah, glob then changes right back.